### PR TITLE
fix #1416 Add lambda subscribe variants with a context

### LINKED ISSUE #1416: subscribe(lambdas...) should have a variant that takes a `Context`
This would have 2 benefits:
 - one intermediate operator less for the `.subscriberContext(Context.of(...)).subscribe(...)` case
 - the `subscribe` would then benefit from `Context`-sensitive customizations (like discard/custom onErrorDropped hooks, etc...)

Right now, for the second point, customizations via `Hooks.onLastOperator` doesn't work because the final subscriber would be replaced. In itself it still doesn't see a `Context`.

Note this is very niche (it only impacts the `LambdaSubscriber#onError` call to `onErrorDropped`, which is for now the only inner use of a hook).

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by smaldini** [comment]
I would rather deprecate the Subscription variant and replace with a Context argument instead

File: reactor-core/src/test/java/reactor/core/publisher/LambdaMonoSubscriberTest.java [hunk 1]
```diff
@@ -19,17 +19,47 @@
...
 import static org.junit.Assert.*;
 
 public class LambdaMonoSubscriberTest {
 
+
+	@Test
+	public void initialContextIsVisibleToUpstream() {
+		AtomicReference<Context> contextRef = new AtomicReference<>();
+
+		Mono.subscriberContext()
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/publisher/LambdaMonoSubscriberTest.java]: I just realized... `deferWithContext` should make the testing easier as well:
```java
var context = Mono.deferWithContext(Mono::just).subscribe(null, null, null, null, Context.of("subscriber", "context"))
```

**Comment by simonbasle** [comment]
@smaldini @bsideup I updated the PR to deprecate the `Consumer<Subscription>` variant and have the `Context` variant only take onNext/onError/onComplete consumers.

**Comment by simonbasle** [comment]
> I would rather deprecate the Subscription variant and replace with a Context argument instead

Let's defer that decision to deprecate for 3.4, this is coming too late (and we have downstream projects that use the method, albeit for tests). 

This PR should focus on adding the `Context` overload, not removing functionality.

However in the last commits I have prepared for such a deprecation (method is not used in tests anymore) and added a TODO. The overload also doesn't take the `Consumer<Subscription>` lambda. There should be no compiler ambiguity there, unless users call the method with `null` arguments (which makes no real sense since this is the last argument and there are overloads with smaller set of inputs).

**Comment by bsideup** [comment]
Out of curiously, what was the primary motivation for adding it?

Just thinking whether focusing on the macro fusion between `subscriberContext` and `subscribe` is better?

**Comment by simonbasle** [comment]
see also https://github.com/reactor/reactor-core/issues/1825 for a future follow-up in the 3.4 timeframe

**Comment by simonbasle** [comment]
> Out of curiously, what was the primary motivation for adding it?
> 
> Just thinking whether focusing on the macro fusion between `subscriberContext` and `subscribe` is better?

lambda subscriber cannot really fuse with the upstream as it is not an operator. having the `Context` as early as this allows internals of the subscriber to also benefit from the `Context` (eg. hooks being set up locally, etc...)

**Pull request merged**